viewing_party/party.py

```diff
@@ -1,23 +1,176 @@
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            break
+
+    return user_data    
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    rating_list = []
+    if not user_data["watched"]:
+        return 0.0
+    for movie in user_data["watched"]:
+        rating_list.append(movie["rating"])
+    average_rating = sum(rating_list) / len(rating_list)
+    return average_rating
+
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+    
+    genre_list = []
+    for movie in user_data["watched"]:
+        genre_list.append(movie["genre"])
+    
+    most_watched_genre_frequency = 0
+    most_watched_genre = genre_list[0]
+    for genre in genre_list:
+        genre_frequency = genre_list.count(genre)
+        if genre_frequency > most_watched_genre_frequency:
+            most_watched_genre_frequency = genre_frequency
+            most_watched_genre = genre
+
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+
+    user_watched_titles =[]
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+    
+    friends_watched_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in friends_watched_titles:
+                friends_watched_titles.append(movie["title"])
+    
+    user_unique_titles = []
+    for title in user_watched_titles:
+        if title not in friends_watched_titles:
+            user_unique_titles.append(title)
+    
+    user_unique_movies = []
+    for title in user_unique_titles:
+        for movie in user_data["watched"]:
+            if movie["title"] == title:
+                user_unique_movies.append(movie)
+
+    return user_unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    
+    user_watched_titles =[]
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+    
+    friends_watched_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in friends_watched_titles:
+                friends_watched_titles.append(movie["title"])
+    
+    friends_unique_titles = []
+    for title in friends_watched_titles:
+        if title not in user_watched_titles:
+            friends_unique_titles.append(title)
+    
+    friends_unique_movies = []
+    for title in friends_unique_titles:
+        for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                if movie["title"] == title and movie not in friends_unique_movies:
+                    friends_unique_movies.append(movie)

```

Comment:
When I see many for loops nested like this, I think about how we could pull values into variables so we don't get deeply indented levels of code. For example, instead of needing two for loops to get into friends' movies, could you create a list of all friends movies before this and then use that list here to save on a for loop?

Also, consider how you could use sets here too because you have for loops using the `in` operator on lists which can quickly lead to quadratic or even cubic time complexity. 

You could create a set of user's watched movie titles and have the inverse logic of get_unique_watched.